chain/network/src/snapshot_hosts/tests.rs

```diff
@@ -200,3 +202,142 @@ async fn test_lru_eviction() {
+            // Now 0 and 2 have been returned once and 1 has been returned twice.
+            // We should go back to 0 and 2 once each, when each peer will have been returned twice.
+            // Then after that it's just a normal looking round robin
+            SelectPeerAction::CallSelect(Some(0)),
+            SelectPeerAction::CallSelect(Some(2)),
+            SelectPeerAction::CallSelect(Some(0)),
+            SelectPeerAction::CallSelect(Some(1)),
+            SelectPeerAction::CallSelect(Some(2)),
+        ],
+    },
+    SelectPeerTest {
+        num_peers: 2,
+        actions: &[
+            SelectPeerAction::CallSelect(None),
+            SelectPeerAction::InsertHosts(&[0]),
+            SelectPeerAction::CallSelect(Some(0)),
+            SelectPeerAction::CallSelect(Some(0)),
+            SelectPeerAction::CallSelect(Some(0)),
+            SelectPeerAction::CallSelect(Some(0)),
+            SelectPeerAction::CallSelect(Some(0)),
+            SelectPeerAction::InsertHosts(&[1]),
+            SelectPeerAction::CallSelect(Some(1)),
+            SelectPeerAction::CallSelect(Some(1)),
+            // calling part_received() should result in correct/acceptable subsequent values
+            // returned by select_host(), but where we forget everything about previous calls
+            SelectPeerAction::PartReceived,
+            SelectPeerAction::CallSelect(Some(0)),
+            SelectPeerAction::CallSelect(Some(1)),
+            SelectPeerAction::CallSelect(Some(0)),
+            SelectPeerAction::CallSelect(Some(1)),
+        ],
+    },
+];
+
+async fn run_select_peer_test(
+    actions: &[SelectPeerAction],
+    peers: &[Arc<SnapshotHostInfo>],
+    sync_hash: &CryptoHash,
+    part_id: &PartId,
+) {
+    let config = Config {
+        snapshot_hosts_cache_size: peers.len() as u32,
+        part_selection_cache_batch_size: 1,
+    };
+    let cache = SnapshotHostsCache::new(config);
+
+    tracing::debug!("start run_select_peer_test");
+
+    for action in actions.iter() {
+        tracing::debug!("run_select_peer_test action {:?}", action);
+        match action {
+            SelectPeerAction::InsertHosts(hosts) => {
+                let mut new_hosts = Vec::new();
+                for h in hosts.iter() {
+                    new_hosts.push(peers[*h].clone());
+                }
+                let (_res, err) = cache.insert(new_hosts).await;
+                assert!(err.is_none());
+            }
+            SelectPeerAction::CallSelect(wanted) => {
+                let peer = cache.select_host(sync_hash, 0, &part_id);
+                let wanted = match wanted {
+                    Some(idx) => Some(&peers[*idx].peer_id),
+                    None => None,
+                };
+                assert!(peer.as_ref() == wanted, "got: {:?} want: {:?}", &peer, &wanted);
+            }
+            SelectPeerAction::PartReceived => {
+                cache.part_received(sync_hash, 0, &part_id);
```

Comment:
Maybe add an assert here that the internal state is actually cleared out when the part is received.